File: check_syntax.js

import fs from 'fs';
import process from 'process';
const content = fs.readFileSync(process.argv[2], 'utf8');
const stack = [];
const lines = content.split('\n');
for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    for (let char of line) {
        if ('{[('.includes(char)) {
            stack.push({ char, line: i + 1 });
        } else if ('}])'.includes(char)) {
            if (stack.length === 0) {
                console.log(`Extra closing ${char} at line ${i + 1}`);
                process.exit(1);
            }
            const top = stack.pop();
            if ((top.char === '{' && char !== '}') ||
                (top.char === '[' && char !== ']') ||
                (top.char === '(' && char !== ')')) {
                console.log(`Mismatched ${top.char} from line ${top.line} with ${char} at line ${i + 1}`);
                process.exit(1);
            }
        }
    }
}
if (stack.length > 0) {
    const top = stack.pop();
    console.log(`Unclosed ${top.char} from line ${top.line}`);
    process.exit(1);
} else {
    console.log("Basic balance check passed");
}

File: check_syntax.py

import sys

def check_syntax(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Very basic check for balanced braces and parentheses in the whole file
    stack = []
    lines = content.split('\n')
    for i, line in enumerate(lines):
        for char in line:
            if char in '{[(':
                stack.append((char, i+1))
            elif char in '}])':
                if not stack:
                    print(f"Extra closing {char} at line {i+1}")
                    return
                opening, line_num = stack.pop()
                if (opening == '{' and char != '}') or \
                   (opening == '[' and char != ']') or \
                   (opening == '(' and char != ')'):
                    print(f"Mismatched {opening} from line {line_num} with {char} at line {i+1}")
                    return
    if stack:
        opening, line_num = stack.pop()
        print(f"Unclosed {opening} from line {line_num}")
    else:
        print("Basic balance check passed")

if __name__ == "__main__":
    check_syntax(sys.argv[1])
